File: src/llm.py
"""LLM provider abstraction for AICrawler."""

import json
import logging
import os
from abc import ABC, abstractmethod

import httpx

logger = logging.getLogger(__name__)


class LLMProvider(ABC):
    """Abstract base class for LLM providers."""

    @abstractmethod
    def generate(self, prompt: str, max_tokens: int = 1024) -> str | None:
        """Generate a response from the LLM."""

    @abstractmethod
    def is_configured(self) -> bool:
        """Check if the provider is properly configured."""


class OllamaProvider(LLMProvider):
    """Ollama local LLM provider."""

    def __init__(self, model: str = "qwen2.5:7b", base_url: str = "http://localhost:11434"):
        self.model = model
        self.base_url = base_url
        self._available: bool | None = None

    def is_configured(self) -> bool:
        if self._available is not None:
            return self._available

        try:
            with httpx.Client(timeout=5.0) as client:
                response = client.get(f"{self.base_url}/api/tags")
                if response.status_code == 200:
                    models = response.json().get("models", [])
                    model_names = [m.get("name", "") for m in models]
                    model_base = self.model.split(":")[0]
                    self._available = any(model_base in name for name in model_names)
                    if not self._available:
                        logger.warning(
                            "Ollama model '%s' not found. Available: %s",
                            self.model,
                            model_names,
                        )
                    return self._available
        except httpx.ConnectError:
            logger.warning("Ollama not running at %s", self.base_url)
            self._available = False
        except Exception as e:
            logger.warning("Error checking Ollama: %s", e)
            self._available = False

        return self._available

    def generate(self, prompt: str, max_tokens: int = 1024) -> str | None:
        try:
            with httpx.Client(timeout=120.0) as client:
                response = client.post(
                    f"{self.base_url}/api/chat",
                    json={
                        "model": self.model,
                        "messages": [{"role": "user", "content": prompt}],
                        "stream": False,
                        "options": {
                            "num_predict": max_tokens,
                            "temperature": 0.3,
                        },
                    },
                )
                response.raise_for_status()
                return response.json()["message"]["content"]
        except Exception as e:
            logger.error("Ollama API error: %s", e)
            return None


class OpenAIProvider(LLMProvider):
    """OpenAI API provider."""

    def __init__(self, model: str = "gpt-4o-mini", api_key_env: str = "OPENAI_API_KEY"):
        self.model = model
        self.api_key = os.environ.get(api_key_env)
        self._client = None

        if self.api_key:
            from openai import OpenAI

            self._client = OpenAI(api_key=self.api_key)
        else:
            logger.debug("OpenAI API key not found in %s", api_key_env)

    def is_configured(self) -> bool:
        return self._client is not None

    def generate(self, prompt: str, max_tokens: int = 1024) -> str | None:
        if not self._client:
            return None

        try:
            response = self._client.chat.completions.create(
                model=self.model,
                messages=[{"role": "user", "content": prompt}],
                max_tokens=max_tokens,
                temperature=0.3,
            )
            return response.choices[0].message.content
        except Exception as e:
            logger.error("OpenAI API error: %s", e)
            return None


def create_provider(config: dict) -> LLMProvider | None:
    """Create LLM provider based on configuration."""
    summ_config = config.get("summarization", {})
    provider_name = summ_config.get("provider", "ollama").lower()

    if provider_name == "ollama":
        model = summ_config.get("model", "qwen2.5:7b")
        base_url = summ_config.get("ollama_url", "http://localhost:11434")
        provider = OllamaProvider(model=model, base_url=base_url)
        if provider.is_configured():
            logger.info("Using Ollama with model: %s", model)
            return provider
        logger.info("Ollama not available, trying OpenAI fallback...")
        provider_name = "openai"

    if provider_name == "openai":
        model = summ_config.get("openai_model", summ_config.get("model", "gpt-4o-mini"))
        api_key_env = summ_config.get("api_key_env", "OPENAI_API_KEY")
        provider = OpenAIProvider(model=model, api_key_env=api_key_env)
        if provider.is_configured():
            logger.info("Using OpenAI with model: %s", model)
            return provider

    logger.error("No LLM provider available. Check Ollama is running or set OPENAI_API_KEY.")
    return None


def parse_json_response(text: str) -> dict | None:
    """Parse a JSON response from an LLM, handling markdown code blocks."""
    text = text.strip()

    # Strip markdown code fences
    if text.startswith("```"):
        lines = text.split("\n")
        end_idx = len(lines) - 1
        for i in range(len(lines) - 1, 0, -1):
            if lines[i].strip() == "```":
                end_idx = i
                break
        text = "\n".join(lines[1:end_idx])

    try:
        return json.loads(text)
    except json.JSONDecodeError as e:
        logger.warning("Failed to parse LLM response as JSON: %s", e)
        logger.debug("Response was: %s", text[:200])
        return None

File: internal/llm/llm.go
package llm

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"log"
	"net/http"
	"os"
	"strings"
	"time"
)

// Provider is the interface for LLM providers.
type Provider interface {
	Generate(ctx context.Context, prompt string, maxTokens int) (string, error)
	IsConfigured() bool
}

// Embedder is the interface for generating embeddings.
type Embedder interface {
	Embed(ctx context.Context, texts []string) ([][]float64, error)
}

// OllamaProvider is a local Ollama LLM provider.
type OllamaProvider struct {
	Model   string
	BaseURL string
	client  *http.Client
}

// NewOllamaProvider creates a new Ollama provider.
func NewOllamaProvider(model, baseURL string) *OllamaProvider {
	return &OllamaProvider{
		Model:   model,
		BaseURL: baseURL,
		client:  &http.Client{Timeout: 120 * time.Second},
	}
}

// IsConfigured checks if Ollama is running and the model is available.
func (o *OllamaProvider) IsConfigured() bool {
	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
	defer cancel()

	req, err := http.NewRequestWithContext(ctx, "GET", o.BaseURL+"/api/tags", nil)
	if err != nil {
		return false
	}

	resp, err := o.client.Do(req)
	if err != nil {
		return false
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return false
	}

	var result struct {
		Models []struct {
			Name string `json:"name"`
		} `json:"models"`
	}
	if err := json.NewDecoder(resp.Body).Decode(&result); err != nil {
		return false
	}

	modelBase := strings.SplitN(o.Model, ":", 2)[0]
	for _, m := range result.Models {
		if strings.Contains(m.Name, modelBase) {
			return true
		}
	}
	log.Printf("Ollama model %q not found", o.Model)
	return false
}

// Generate sends a prompt to Ollama and returns the response.
func (o *OllamaProvider) Generate(ctx context.Context, prompt string, maxTokens int) (string, error) {
	body := map[string]any{
		"model": o.Model,
		"messages": []map[string]string{
			{"role": "user", "content": prompt},
		},
		"stream": false,
		"options": map[string]any{
			"num_predict":  maxTokens,
			"temperature": 0.3,
		},
	}

	data, err := json.Marshal(body)
	if err != nil {
		return "", fmt.Errorf("marshaling request: %w", err)
	}

	req, err := http.NewRequestWithContext(ctx, "POST", o.BaseURL+"/api/chat", bytes.NewReader(data))
	if err != nil {
		return "", fmt.Errorf("creating request: %w", err)
	}
	req.Header.Set("Content-Type", "application/json")

	resp, err := o.client.Do(req)
	if err != nil {
		return "", fmt.Errorf("ollama API error: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		respBody, _ := io.ReadAll(resp.Body)
		return "", fmt.Errorf("ollama API returned %d: %s", resp.StatusCode, string(respBody))
	}

	var result struct {
		Message struct {
			Content string `json:"content"`
		} `json:"message"`
	}
	if err := json.NewDecoder(resp.Body).Decode(&result); err != nil {
		return "", fmt.Errorf("decoding response: %w", err)
	}

	return result.Message.Content, nil
}

// OllamaEmbedder generates embeddings via the Ollama API.
type OllamaEmbedder struct {
	Model   string
	BaseURL string
	client  *http.Client
}

// NewOllamaEmbedder creates a new Ollama embedder.
func NewOllamaEmbedder(model, baseURL string) *OllamaEmbedder {
	return &OllamaEmbedder{
		Model:   model,
		BaseURL: baseURL,
		client:  &http.Client{Timeout: 120 * time.Second},
	}
}

// Embed generates embeddings for the given texts.
func (e *OllamaEmbedder) Embed(ctx context.Context, texts []string) ([][]float64, error) {
	body := map[string]any{
		"model": e.Model,
		"input": texts,
	}

	data, err := json.Marshal(body)
	if err != nil {
		return nil, fmt.Errorf("marshaling request: %w", err)
	}

	req, err := http.NewRequestWithContext(ctx, "POST", e.BaseURL+"/api/embed", bytes.NewReader(data))
	if err != nil {
		return nil, fmt.Errorf("creating request: %w", err)
	}
	req.Header.Set("Content-Type", "application/json")

	resp, err := e.client.Do(req)
	if err != nil {
		return nil, fmt.Errorf("ollama embed error: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		respBody, _ := io.ReadAll(resp.Body)
		return nil, fmt.Errorf("ollama embed returned %d: %s", resp.StatusCode, string(respBody))
	}

	var result struct {
		Embeddings [][]float64 `json:"embeddings"`
	}
	if err := json.NewDecoder(resp.Body).Decode(&result); err != nil {
		return nil, fmt.Errorf("decoding embeddings: %w", err)
	}

	return result.Embeddings, nil
}

// OpenAIProvider is an OpenAI API provider.
type OpenAIProvider struct {
	Model  string
	APIKey string
	client *http.Client
}

// NewOpenAIProvider creates a new OpenAI provider.
func NewOpenAIProvider(model, apiKeyEnv string) *OpenAIProvider {
	return &OpenAIProvider{
		Model:  model,
		APIKey: os.Getenv(apiKeyEnv),
		client: &http.Client{Timeout: 120 * time.Second},
	}
}

// IsConfigured checks if the API key is set.
func (o *OpenAIProvider) IsConfigured() bool {
	return o.APIKey != ""
}

// Generate sends a prompt to OpenAI and returns the response.
func (o *OpenAIProvider) Generate(ctx context.Context, prompt string, maxTokens int) (string, error) {
	if o.APIKey == "" {
		return "", fmt.Errorf("OpenAI API key not configured")
	}

	body := map[string]any{
		"model": o.Model,
		"messages": []map[string]string{
			{"role": "user", "content": prompt},
		},
		"max_tokens":  maxTokens,
		"temperature": 0.3,
	}

	data, err := json.Marshal(body)
	if err != nil {
		return "", fmt.Errorf("marshaling request: %w", err)
	}

	req, err := http.NewRequestWithContext(ctx, "POST", "https://api.openai.com/v1/chat/completions", bytes.NewReader(data))
	if err != nil {
		return "", fmt.Errorf("creating request: %w", err)
	}
	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("Authorization", "Bearer "+o.APIKey)

	resp, err := o.client.Do(req)
	if err != nil {
		return "", fmt.Errorf("OpenAI API error: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		respBody, _ := io.ReadAll(resp.Body)
		return "", fmt.Errorf("OpenAI API returned %d: %s", resp.StatusCode, string(respBody))
	}

	var result struct {
		Choices []struct {
			Message struct {
				Content string `json:"content"`
			} `json:"message"`
		} `json:"choices"`
	}
	if err := json.NewDecoder(resp.Body).Decode(&result); err != nil {
		return "", fmt.Errorf("decoding response: %w", err)
	}

	if len(result.Choices) == 0 {
		return "", fmt.Errorf("no choices in OpenAI response")
	}

	return result.Choices[0].Message.Content, nil
}

// CreateProvider creates an LLM provider based on configuration.
func CreateProvider(provider, model, ollamaURL, openaiModel, apiKeyEnv string) Provider {
	if strings.ToLower(provider) == "ollama" {
		p := NewOllamaProvider(model, ollamaURL)
		if p.IsConfigured() {
			log.Printf("Using Ollama with model: %s", model)
			return p
		}
		log.Println("Ollama not available, trying OpenAI fallback...")
	}

	p := NewOpenAIProvider(openaiModel, apiKeyEnv)
	if p.IsConfigured() {
		log.Printf("Using OpenAI with model: %s", openaiModel)
		return p
	}

	log.Println("No LLM provider available. Check Ollama is running or set OPENAI_API_KEY.")
	return nil
}
